### ISSUE TITLE
Network policy / SSH restrictions 

### ISSUE THREAD

**Issue by jhprinz** [markovmodel/adaptive-sampling#13]
This is something that @shantenujha and @andre-merzky and @vivek-bala might help me with.

I was trying to set EnsembleTK up on allegro to test locally. Ultimately we would like to use it here, too. Not only on Titan. The current problem is that the nodes on Allegro seem not to be able to create an outward connection, even with SSH and I am still looking for machines the nodes could connect to.

The main problem I see is to place the MongoDB somewhere, where we can have access to it while the simulation is running. EnsTK should work as one big standalone task, but I think the control from an external DB would be preferred. I could imagine that this might run on the head node or we create a tunnel on the head node if possible. 

I was just wondering, what experience you guys have with this problem. It seems that the worker approach is used a lot in other fields, but most HPC clusters are not meant to run tasks to be controlled from outside.

**Comment by andre-merzky** [comment]
radical.pilot faces the same problem on several other clusters, too -- but it actually should create an ssh tunnel on the fly in those cases.  The resource config can specify the following keys for that:

* `tunnel_bind_device`: the network device to be used for the tunnel.  This should be the device which can connect to the head node via ssh
* `forward_tunnel_endpoint`: the ip number to connect to.  This would normally be the address of the head node.

We do some magic in the bootstrapper to accommodate different network configurations -- if you are interested in the details, feel free to have a look at 

  * https://github.com/radical-cybertools/radical.pilot/blob/devel/src/radical/pilot/bootstrapper/bootstrap_1.sh#L1286

Please let us know if that makes sense, and if we can help digging out the respective config parameters.

**Comment by jhprinz** [comment]
Thanks. This helps a lot, but is way more advanced what I quickly hacked together. So let me just rephrase the minimal requirements:

There needs to be at least one node/machine that can connect to the head or anywhere outside to establish the forward ssh tunnel to connect to the mondosb server. 

I assume that most clusters have some way to allow for this or a special machine/hub to connect to.

**Comment by andre-merzky** [comment]
> There needs to be at least one node/machine that can connect to the head or anywhere outside to establish the forward ssh tunnel to connect to the mongodb server.

Indeed.  More precisely: the node on which the pilot agent is running needs to be at most one ssh-hop away from some node which can connect to the MongoDB.  All other compute nodes need to be able to communicate directly with the node the pilot agent is running on.

**Comment by jhprinz** [comment]
Thanks, yes, that is what I thought. IN most cases I will just place the pilot in the queue and it does all the acquisition of the nodes and distributes the work. 

Do you have experience with running the DB next to the pilot job? Like starting the DB on one node and the start using `python run_radical.py`. There might be enough ressource to run both on the same node. 

I think that is not the optimum but could be useful to have.

**Comment by andre-merzky** [comment]
Running the DB on some node should *in principle* work -- but keep in mind that the address of the DB needs to be known to the client application *before* it submits the pilot (otherwise we are playing chicken-and-egg).  Also, the client application needs to be able to contact the MongoDB, and calling into the compute nodes is usually more difficult than calling out...

**Comment by jhprinz** [comment]
okay, so when setting up mongodb we need to pass the URL to the pilot before it boots up so everyone knows. We should try to avoid that. I already contacted our cluster personnel to figure out how to do that.
